### ISSUE TITLE
Additional AWS managed resources

### ISSUE THREAD

**Issue by infinitecompute** [crossplane-contrib/provider-aws#149]
### What problem are you facing?

We are gathering community feedback to help us prioritize development of additional AWS managed services and maturing of existing service implementations.
* What are the most important AWS services for you?  Please share your service list in the comments.
* What are your use cases?  This will help us understand how to best support your situation.
* Would you be interested in contributing?  If so, in which capacity?  This could take the form of usage and early feedback, code contributions, and improved documentation.

Crossplane currently supports [58+ AWS API types](https://doc.crds.dev/github.com/crossplane/provider-aws), see https://doc.crds.dev/github.com/crossplane/provider-aws.

![image](https://user-images.githubusercontent.com/284840/103931999-0d391400-50d6-11eb-8055-11857b46b687.png)

Please drop us a comment with a list of the most important AWS services for your use cases.

### How could Crossplane help solve your problem?

We will be prioritizing updates and additional services in Crossplane based on feedback. 

### Related Issues

* [x] VCP, Subnet, SecurityGroup and InternetGateway at v1beta1 #182 
* [x] IAM from v1alpha3 -> v1beta1 #100 
* [x] DBSubnet Group -> v1beta1 #165
* [x] DynamoDB Table #147
* [x] More IAM types - #151
   * IAMRole, IAMRolePolicyAttachment -> v1beta1 #141 
   * IAMUser #194, IAMUserPolicyAttachment #196 
   * IAMGroup, IAMGroupUserMembership, IAMGroupPolicyAttachment #249 
   * IAMUserPolicyAttachment refers to IAMPolicy #231, PR #257
* [x] DNS: Route 53 #172
* [x] Certificate Manager #171
* [x] SQS #170
* [x] ELB #180
* [x] SNS #175
* [x] RouteTable to v1beta1 #145
* [x] S3 bucket API types and controllers to v1beta1 quality [#331](https://github.com/crossplane/provider-aws/pull/331)
* [x] ElastiCache #168, PR #189 
* [x] ECR support #307
* [x] S3 Bucket Policy to v1beta1 [#391](https://github.com/crossplane/provider-aws/pull/391)
* [x] IAM User Access Key v1alpha1 [#403](https://github.com/crossplane/provider-aws/pull/403)
* [x] Add API Gateway as a managed resource #235
* [x] ACK code generation of native Crossplane AWS resources and `provider-aws` [see note below](https://github.com/crossplane/provider-aws/issues/149#issuecomment-718208201)
   * see [Crossplane Community Day #2](https://www.youtube.com/playlist?list=PL510POnNVaaZx1QDIn_-77igM1COuHaHV) talk by @jaypipes (ACK lead) and @muvaf (Crossplane Maintainer) on [Accelerating Crossplane AWS Provider coverage with ACK code generation](https://www.youtube.com/watch?v=4sM3tmzN2ZI&list=PL510POnNVaaZx1QDIn_-77igM1COuHaHV&index=9) that shows ACK generated Crossplane provider resources.

#### Up Next
* [ ] Add AWS ElasticSearch Service as managed resource #238
* [ ] Add AWS EMR Cluster as a managed resource #239
* [x] Lambda #234 
* [x] DocumentDB (with MongoDB compatibility) #268 
* [x] Cache Subnet Group #169, #95
* [x] Add AWS Lambda Function as a custom resource #234
* [x] Add CloudFront as a managed resource #236
* [ ] Add Auto Scaling Group as a managed resource #237

#### Parked for Design Review

* Kinesis

**Comment by muvaf** [comment]
Elasticache provisioned by Crossplane cannot live in a VPC right now due to missing managed resource `CacheSubnetGroup` right now. It's a fairly simple resource as it's basically a logical grouping of subnets. We could consider implementing it. See https://github.com/crossplane/provider-aws/issues/95 for details.

**Comment by infinitecompute** [comment]
Noted, I'll add this dependency to the prioritization Muvaffak. thanks!

**Comment by muvaf** [comment]
> RDS from alpha to beta

@infinitecompute RDS is already `v1beta1`

**Comment by janwillies** [comment]
We are a user of `DocumentDB`, it would be great to see this supported: https://github.com/crossplane/provider-aws/issues/268

**Comment by nicorikken** [comment]
We are looking into [ECR](https://aws.amazon.com/ecr/) support. Our use-case (as discussed on Slack):

> Currently we have deployed our own container registry within the Kubernetes cluster. We'd like to move to ECR to have a managed solution and also increase security by having seperate registries's. However, this means we need to migrate in the order of a 100 container registries's to separate ECR registries. But apart from that initial migration, teams ideally can create container registries just like they can create github repo's. Most are familiar with Kubernetes, but less so with AWS. So having ECR support in Crossplane could ease the creation of cluster-specific registries for teams, rather than having to learn the AWS UI. Also team-specific resources are now all defined in Kubernetes, so it would be nice to tie the ECR creation in to that templating (managed through ArgoCD Gitops).
> To be honest, I'm not yet sure what would be easier: 1) clear team instructions how to use the AWS Console, 2) Some Cloudformation solution, 3) Managing from Kubernetes through Crossplane.

I'm still not sure if Crossplane is the right solution for our use-case, that might take until October this year. If it does, we might contribute code to implement this feature.

**Comment by prasek** [comment]
> We are a user of DocumentDB, it would be great to see this supported: #268
@janwillies added to the list ☝️ 

> We are looking into ECR support. Our use-case (as discussed on Slack) ...
@nicorikken - looks like @muvaf created #307 for this and it's added to the list

Also note that we're looking at code generation options for the AWS provider (e.g. crossplane/crossplane#262) so we'll be getting support for a lot more cloud services shortly. 🚀

**Comment by prasek** [comment]
> We are looking into ECR support. Our use-case (as discussed on Slack):

@nicorikken - @enderv added ECR support in https://github.com/crossplane/provider-aws/pull/337, let us know if you run into any issues, thx!

https://doc.crds.dev/github.com/crossplane/provider-aws/ecr.aws.crossplane.io/Repository/v1alpha1

**Comment by prasek** [comment]
@krishchow thanks for your contributions adding support for:
* S3 Bucket to v1beta1 [#331](https://github.com/crossplane/provider-aws/pull/331)
* S3 Bucket Policy support [#289](https://github.com/crossplane/provider-aws/pull/289)
* Referencer for SubnetGroup AWS ElasticCache [#314](https://github.com/crossplane/provider-aws/pull/314)
* Add ARN to AtProvider for SNS Topic [#348](https://github.com/crossplane/provider-aws/pull/348)

plus a few more on the way:
* S3 Bucket Policy to v1beta1 [#391](https://github.com/crossplane/provider-aws/pull/391)
* IAM User Access Key v1alpha1 [#403](https://github.com/crossplane/provider-aws/pull/403)

nice work! 🚀

**Comment by prasek** [comment]
Excited to see the joint work between @jaypipes, @muvaf, and @kasey to adapt the AWS ACK codegen pipeline to emit a native Crossplane `provider-aws`, so you can compose even more AWS cloud services using Crossplane:

https://twitter.com/jaypipes/status/1320714029917229064
> Hi Jan! Yep, @muvaf  has been digging in to the ACK code generator and running with a proof-of-concept hack that I whipped up a few weeks ago. I'm looking forward to reviewing his PRs this week and going further down the road of code re-use and sharing.

To follow along checkout this PR: https://github.com/aws/aws-controllers-k8s/pull/444 🚀

**Comment by janwillies** [comment]
#407 (fargate) and #405 (secrets manager)

**Comment by lvalerio** [comment]
What are the most important AWS services for you?
#234 ( Lambda Function )

What are your use cases? 
Event-driven programming. Need to deploy a Lambda function to run on events generated from an s3 bucket

Would you be interested in contributing? If so, in which capacity? 
Code contributions and improved documentation

**Comment by muvaf** [comment]
Hi @lvalerio If you're interested in contributing #234 , you can take a look at our [Provider Development Guide](https://crossplane.io/docs/v0.14/contributing/provider_development_guide.html) to get started. Feel free to join our Slack, too. Happy to help you along 🙂

**Comment by muvaf** [comment]
Here is the list of resources that we'll implement using AWS code generation pipeline:

* [ ] https://github.com/crossplane/provider-aws/issues/451 SNS missing resources
* [ ] https://github.com/crossplane/provider-aws/issues/509 KMS
* [ ] https://github.com/crossplane/provider-aws/issues/510 Amazon MQ
* [ ] https://github.com/crossplane/provider-aws/issues/511 RDS missing resources
* [ ] https://github.com/crossplane/provider-aws/issues/512 Route53 missing resources
* [ ] https://github.com/crossplane/provider-aws/issues/513 SQS missing resources
* [ ] https://github.com/crossplane/provider-aws/issues/458 SageMaker

I believe this issue has been a long-running one and we may never be able to close it if we keep adding. So, I plan to close it after this list of issues are closed.

**Comment by dradetsky** [comment]
One service which my org will need & which is not yet mentioned here is EventBridge. Unfortunately, I'm not super-familiar with how we're currently using EventBridge, so I can't give a ton of useful info on our use-case (I should probably find out, huh?) but I believe it's basically

- There are several serverless apps which have their own EventBus's.
- Some serverless apps use EventBridge rules as a kind of crond-for-lambda.

Basically, there will probably be more of these kinds of rules added, and we'd like to provide a (much better) interface for app devs to define & use these rules.

**Comment by DWSR** [comment]
Hey all, we're investigating how to create an SQS-triggered Lambda function using Crossplane. It _appears_ that creating the event source mapping between the SQS Queue and the Lambda function is currently unsupported in Crossplane based on [prior art from Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping).

This use case is super important to us because we need to be able to use Lambda to transform messages placed on an SQS queue (we're unable to use any other offering for `$reasons`).

**Comment by mikebollandajw** [comment]
We are also looking for Cloudwatch Event Rules / EventBridge to SNS (https://github.com/crossplane/provider-aws/issues/1150)

**Comment by jrake-revelant** [comment]
Any chance the code-generation will be picked up again? It looks like only one of https://github.com/crossplane-contrib/provider-aws/issues/149#issuecomment-768903776 got worked on. It would be great to know if we should count on more work of ACK being re-used. Thanks!

**Comment by haarchri** [comment]
please Open dedicated issues for new resources - we will close the ticket here

**Comment by xchan** [comment]
> Here is the list of resources that we'll implement using AWS code generation pipeline:
> 
> * [ ]  [Generate SNS services using ACK #451](https://github.com/crossplane-contrib/provider-aws/issues/451) SNS missing resources
> * [ ]  [Implement all KMS services #509](https://github.com/crossplane-contrib/provider-aws/issues/509) KMS
> * [ ]  [Implement Amazon MQ service #510](https://github.com/crossplane-contrib/provider-aws/issues/510) Amazon MQ
> * [ ]  [Implement missing RDS resources #511](https://github.com/crossplane-contrib/provider-aws/issues/511) RDS missing resources
> * [ ]  [Implement missing Route53 resources #512](https://github.com/crossplane-contrib/provider-aws/issues/512) Route53 missing resources
> * [ ]  [Implement SQS services #513](https://github.com/crossplane-contrib/provider-aws/issues/513) SQS missing resources
> * [ ]  [Generate SageMaker Services using ACK #458](https://github.com/crossplane-contrib/provider-aws/issues/458) SageMaker
> 
> I believe this issue has been a long-running one and we may never be able to close it if we keep adding. So, I plan to close it after this list of issues are closed.

Any progress on SageMaker Services?
